### PR TITLE
Add support for connections to multiple deployments

### PR DESCRIPTION
**Pull request by code-asher**
This is broadly three parts, as described by the commits, but the main gist is that the SSH blocks are now scoped (with the deployment URL in the host name), and the individual config files referenced in those config blocks are also scoped (with the deployment URL in the directory), and the connection process has been decoupled from the credentials in VS Code's memory and coupled to the credentials in the associated config directory instead.  More context in each commit message; it might even make sense to review by commit.

Existing hosts should not be affected, only new connections will be scoped by deployment.  Existing hosts (ones without the deployment in the host) will continue using the old config block and directories.  But it will not be possible to add new hosts using the old style; new connections will always be scoped. 

I should also clarify that this specifically is for supporting the "Recents" menu that may contain connections from multiple deployments and for reconnections (say you connect and then log into a different deployment while the other connection is still up).  It does not let you log into multiple deployments and see all the workspaces from all of them at once.

In other words it closes https://github.com/coder/vscode-coder/issues/173 but **not** https://github.com/coder/vscode-coder/issues/102.

### CODE REVIEW

File: src/util.test.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { it, expect } from "vitest"
+import { toSafeHost } from "./util"
+
+it("escapes url host", async () => {
+  expect(toSafeHost("https://foobar:8080")).toBe("foobar")
+  expect(toSafeHost("https://ほげ")).toBe("xn--18j4d")
```
Comment (buenos-nachos) [inline src/util.test.ts:63]: TIL that "hoge" is the Japanese equivalent of "foo"

File: src/extension.ts [hunk 2]
```diff
@@ -117,9 +119,13 @@ export async function activate(ctx: vscode.ExtensionContext): Promise<void> {
         // prompted to sign in again, so we do not need to ensure it is set.
         const token = params.get("token")
         if (token) {
+          restClient.setSessionToken(token)
           await storage.setSessionToken(token)
         }
 
+        // Store on disk to be used by the cli.
+        await storage.configureCli(toSafeHost(url), url, token)
```
Comment (buenos-nachos) [inline src/extension.ts:127]: I've noticed that all the calls to `storage.configureCli` have also required a call to `toSafeHost`. Do you think that `configureCli` could call `toSafeHost` internally?

File: src/storage.ts [hunk 3]
```diff
@@ -109,16 +105,20 @@ export class Storage {
...
+   * the old deployment-unaware path instead.
+   *
    * If there is already a working binary and it matches the server version,
    * return that, skipping the download.  If it does not match but downloads are
    * disabled, return whatever we have and log a warning.  Otherwise throw if
    * unable to download a working binary, whether because of network issues or
    * downloads being disabled.
    */
-  public async fetchBinary(restClient: Api): Promise<string> {
+  public async fetchBinary(restClient: Api, label: string | string): Promise<string> {
```
Comment (buenos-nachos) [inline src/storage.ts]: Typo? `string | string` collapses down to `string`

**Review by Parkreiner — APPROVED** [review]
Looks good, and everything seems to make sense (I followed your recommendation and went through things commit-by-commit)

I'm gonna want to review more of how this codebase is set up, though. All the individual function calls seemed to make sense, but I feel like there's probably some finer nuances that I'm missing (especially with the VSCode APIs)

File: src/extension.ts (hunk 2 above)
Comment (code-asher) [reply src/extension.ts:127]: That is a good question.  My thinking is that this storage interface is taking a directory name (or label as I am calling it), so from that perspective it does not make too much sense for it to be aware of URLs and how to encode them.

So, for similar reasons that, say, `fs.writeFile(fileName)` would not automatically encode the file name (er, at least, pretty sure it does not).

Similar to how `SSHConfig` takes a label rather than a URL as well.

File: src/storage.ts (hunk 3 above)
Comment (code-asher) [reply src/storage.ts]: Whoops good catch lol will fix

You can pass a string OR A STRING!  DOUBLE STRING

File: src/extension.ts (hunk 2 above)
Comment (code-asher) [reply src/extension.ts:127]: Probably it will always be a URL, but in the future there could be tests doing things like `storage.configureCli("temp-dir")` or some such.  But you could argue that the tests should just use URLs too, and the storage abstraction should change to take URLs rather than generic labels (and `SSHConfig` as well).

But, I am partial to the generic label abstraction personally, maybe because it feels less coupled/simpler to me.

I would have tests already except we need to switch away from vitest back to the VS Code testing framework.  :sob:

**Comment by code-asher** [comment]
> I'm gonna want to review more of how this codebase is set up, though. All the individual function calls seemed to make sense, but I feel like there's probably some finer nuances that I'm missing (especially with the VSCode APIs)

I feel that, it is really tough to do code review in an unfamiliar repository.  I appreciate your reviews all the same though!
